source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.cc

```diff
@@ -32,68 +36,219 @@ CpuTimes LinuxCpuStatsReader::getCpuTimes() {
   const std::string target = "cpu  ";
   if (!cpu_stats_file || buffer != target) {
     ENVOY_LOG_MISC(error, "Unexpected format in linux cpu stats file {}", cpu_stats_filename_);
-    return {false, 0, 0};
+    return {false, false, 0, 0, 0};
   }
 
   std::array<uint64_t, NUMBER_OF_CPU_TIMES_TO_PARSE> times;
   for (uint64_t time, i = 0; i < NUMBER_OF_CPU_TIMES_TO_PARSE; ++i) {
     cpu_stats_file >> time;
     if (!cpu_stats_file) {
       ENVOY_LOG_MISC(error, "Unexpected format in linux cpu stats file {}", cpu_stats_filename_);
-      return {false, 0, 0};
+      return {false, false, 0, 0, 0};
     }
     times[i] = time;
   }
 
   uint64_t work_time, total_time;
   work_time = times[0] + times[1] + times[2]; // user + nice + system
   total_time = work_time + times[3];          // idle
-  return {true, static_cast<double>(work_time), total_time};
+  return {true, false, static_cast<double>(work_time), total_time, 0};
 }
 
 LinuxContainerCpuStatsReader::LinuxContainerCpuStatsReader(
     TimeSource& time_source, const std::string& linux_cgroup_cpu_allocated_file,
-    const std::string& linux_cgroup_cpu_times_file)
+    const std::string& linux_cgroup_cpu_times_file, const std::string& cgroupv2_cpu_stat_file,
+    const std::string& cgroupv2_cpu_max_file, const std::string& cgroupv2_cpu_effective_file)
     : time_source_(time_source), linux_cgroup_cpu_allocated_file_(linux_cgroup_cpu_allocated_file),
-      linux_cgroup_cpu_times_file_(linux_cgroup_cpu_times_file) {}
+      linux_cgroup_cpu_times_file_(linux_cgroup_cpu_times_file),
+      cgroupv2_cpu_stat_file_(cgroupv2_cpu_stat_file),
+      cgroupv2_cpu_max_file_(cgroupv2_cpu_max_file),
+      cgroupv2_cpu_effective_file_(cgroupv2_cpu_effective_file) {}
 
 CpuTimes LinuxContainerCpuStatsReader::getCpuTimes() {
+
+  // Prefer cgroup v2 files if they exist, else fall back to cgroup v1
+  std::ifstream v2_stat_file(cgroupv2_cpu_stat_file_);
+  std::ifstream v2_max_file(cgroupv2_cpu_max_file_);
+  std::ifstream v2_effective_file(cgroupv2_cpu_effective_file_);
+
```

Comment:
We could do all of this detection in the factory that decides which of the implementations to choose vs always trying cgroupv2 files on a system using cgroupv1 inefficiently

This would reduce the size of this function significantly.